### ISSUE TITLE
Fork crashes during interactive rebase with exec

### ISSUE THREAD

**Issue by nsh-ableton** [ForkIssues/Tracker#1682]
A frequent part of my workflow is doing a "check that all commits build" on the branch I'm working on. I start this with a command line like this:

```bash
$ git rebase -ix "build-command " origin/main
```

This then gives me a rebase todo that looks like this:
```
pick 541addd5d9cb Some commit message
exec build-command
pick 5e5d5e9e78b4 Another commit message
exec build-command
pick 5fb751b887f5 Yet another commit message
exec build-command
```

`build-command` returns a non-zero value if building fails, so then I can fix whatever's wrong, amend the commit, and proceed with `git rebase --continue`.

Fork doesn't support this, as far as I know — which is fine! — but the problem is that it also just gets confused and isn't usable while such a rebase is in progress. I find that sad, because I'd love to be able to use Fork for managing the amending of the commit. Unfortunately, the messaging in the program is just confused, seemingly unclear about what kind of process is going on. It seems so confused, in fact, that I have to _avoid_ activating the Fork window, because sometimes it seems to do something weird to the state of the repo at that point and I have to abort and restart the rebase.

It would be wonderful if Fork would just not get confused about this, so I could use all its lovely features while in the midst of this regular part of my workflow.

Thanks!

**Comment by nsh-ableton** [comment]
It seems like Fork is handling these situations less gracefully than it used to. Nowadays, if I accidentally switch to Fork while performing an interactive rebase with `--exec`, Fork immediately crashes. I'm attaching a crash report, in case that's helpful:

[Fork Crash Report.txt](https://github.com/fork-dev/Tracker/files/13887687/Fork.Crash.Report.txt)

**Comment by nsh-ableton** [comment]
@DanPristupov I'd be happy to run a build with debug symbols included if that would help in tracking down the bug. This is very easy for me to reproduce.

**Comment by DanPristupov** [comment]
Crash report is enough, thanks.

1. Can you send me the `.git/rebase-merge/done` file to support@fork.dev? Fork crashes parsing list of SHAs in it.
2. Fork can only continue IR when it was started by Fork. Now it crashes, but I'm not sure if there is a better way. Blocking Fork UI doesn't seem like a better solution. There's no data to neither figure out the IR state nor proceed. Running any rebase command from Fork (like `rebase continue`) will likely cause a wrong result.

**Comment by DanPristupov** [comment]
I can't reproduce the problem myself.

```
git rebase -ix "false " other-branch
Executing: false
warning: execution failed: false
You can fix the problem, and then run

  git rebase --continue
```

Then I open Fork and it doesn't crash. As I said above, the reason of the crash in your case is an unexpected `.git/rebase-merge/done` content.

**Comment by nsh-ableton** [comment]
I've sent you the file :)

**Comment by DanPristupov** [comment]
OK, I've reproduced that.

Unfortunately, what I said is pretty much correct. This is an unsupported flow and handling it
1. requires large IR rework
2. will benefit only a few (or even less ;) ) users, but that's fine :).

We are in the middle of a multi-months task I cannot make *realistic* promises. I'd like to help you, but right now I can't. If the problem still bothers you, can you remind me (maybe even several times) about it in the future, please?

**Comment by nsh-ableton** [comment]
Thanks for looking into it! And sure, I'll be glad to remind you about it in the future. Good luck with your big project!

**Comment by joevt** [comment]
> A frequent part of my workflow is doing a "check that all commits build" on the branch I'm working on. I start this with a command line like this:

I did not think to use interactive rebase to perform this task. This is the method I use:

loop through all commits (starting with the earliest), checkout the commit, perform the build, and output a pass or fail for that commit. I can leave this overnight (since I have a few hundred commits not in upstream), and come back in the morning to see which commits I need to edit with the Fork.app UI.

This method leaves SHA's unchanged unless I need to do an edit.

**Comment by nsh-ableton** [comment]
> loop through all commits (starting with the earliest), checkout the commit, perform the build, and output a pass or fail for that commit. I can leave this overnight (since I have a few hundred commits not in upstream), and come back in the morning to see which commits I need to edit with the Fork.app UI.

This process is exactly what the interactive rebase with `--exec` does 😄

**Comment by joevt** [comment]
> This process is exactly what the interactive rebase with `--exec` does 😄

Yes, but `rebase` changes SHA values even if there are no changes? Doesn't that unnecessarily grow the git history? After a rebase, the original commits before the rebase still exist in the `reflog`. Though that's temporary (90 days?) so I guess it doesn't matter.

**Comment by nsh-ableton** [comment]
> Yes, but `rebase` changes SHA values even if there are no changes? Doesn't that unnecessarily grow the git history? After a rebase, the original commits before the rebase still exist in the `reflog`. Though that's temporary (90 days?) so I guess it doesn't matter.

That's part of what I love about the reflog — I can always retrieve something from the past if I accidentally left it behind.
